src/common/Cryptography/AES.cpp

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright (C) 2008-2019 TrinityCore <https://www.trinitycore.org/>
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation; either version 2 of the License, or (at your
+ * option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "AES.h"
+#include "Errors.h"
+#include <array>
+#include <openssl/evp.h>
+#include <openssl/rand.h>
+
+static constexpr size_t BLOCK_SIZE_BYTES = 16;
+static constexpr size_t KEY_SIZE_BYTES   = 16;
+using IV = std::array<uint8, BLOCK_SIZE_BYTES>;
+using Key = std::unique_ptr<uint8[]>;
+
+static Key ToKey(BigNumber const& bn)
+{
+    ASSERT(bn.GetNumBytes() <= KEY_SIZE_BYTES);
+    return bn.AsByteArray(KEY_SIZE_BYTES);
+}
+
+// PKCS#7 padding
+static void AddPadding(std::vector<uint8>& data)
+{
+    uint8 p = BLOCK_SIZE_BYTES - (data.size() % BLOCK_SIZE_BYTES);
+    data.insert(data.end(), p, p);
+}
+
+static void RemovePadding(std::vector<uint8>& data)
+{
+    ASSERT(data.back() <= data.size());
+    data.resize(data.size() - data.back());
+}
+
+static IV GenerateIV()
+{
+    IV iv;
+    int status = RAND_bytes(iv.data(), BLOCK_SIZE_BYTES);
+    ASSERT(status == 1);
+    return iv;
+}
+
+static void AddIV(std::vector<uint8>& data, IV const& iv)
+{
+    data.insert(data.end(), iv.begin(), iv.end());
+}
+
+static IV RemoveIV(std::vector<uint8>& data)
+{
+    ASSERT(data.size() >= BLOCK_SIZE_BYTES);
+    IV iv;
+    for (size_t i = 1; i <= BLOCK_SIZE_BYTES; ++i)
+    {
+        iv[BLOCK_SIZE_BYTES - i] = data.back();
+        data.pop_back();
+    }
+    return iv;
+}
+
+/*static*/ void AES::Encrypt(std::vector<uint8>& data, BigNumber const& bnKey)
+{
+    // add padding
+    AddPadding(data);
```

Comment:
You don't need to pad your input, OpenSSL EVP has padding enabled by default (PKCS#7 too).

I would also use `_ex` variants of `Init` and `Final`.

Reply:
i know, but i wanted to know in advance how much space i would need to allocate for the output (if i pad manually, input size should equal output size)

Reply:
And if you don't pad manually then output size is just input.length + block_size - input.length() % block_size.